nos/models/llama2_chat.py

```diff
@@ -0,0 +1,139 @@
+import time
+from dataclasses import dataclass
+from threading import Thread
+from typing import Iterable, List, Optional, Tuple
+
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer, TextIteratorStreamer
+
+from nos import hub
+from nos.common import TaskType
+from nos.hub import HuggingFaceHubConfig, hf_login
+
+
+@dataclass(frozen=True)
+class Llama2ChatConfig(HuggingFaceHubConfig):
+    """Llama2 chat model configuration."""
+
+    max_new_tokens: int = 2048
+    """Maximum number of tokens to generate."""
+
+    max_input_token_length: int = 4096
+    """Maximum number of tokens in the input."""
+
+    compute_dtype: str = "float16"
+    """Compute type for the model."""
+
+    needs_auth: bool = False
+    """Whether the model needs authentication."""
+
+
+class Llama2Chat:
+    configs = {
+        "meta-llama/Llama-2-7b-chat-hf": Llama2ChatConfig(
+            model_name="meta-llama/Llama-2-7b-chat-hf",
+            compute_dtype="float16",
+            needs_auth=True,
+        ),
+        "HuggingFaceH4/zephyr-7b-beta": Llama2ChatConfig(
+            model_name="HuggingFaceH4/zephyr-7b-beta",
+            compute_dtype="float16",
+        ),
+        "HuggingFaceH4/tiny-random-LlamaForCausalLM": Llama2ChatConfig(
+            model_name="HuggingFaceH4/tiny-random-LlamaForCausalLM",
+            compute_dtype="float16",
+        ),
+    }
+
+    def __init__(self, model_name: str = "HuggingFaceH4/zephyr-7b-beta"):
+        from nos.logging import logger
+
+        try:
+            self.cfg = Llama2Chat.configs[model_name]
+        except KeyError:
+            raise ValueError(f"Invalid model_name: {model_name}, available models: {Llama2ChatConfig.configs.keys()}")
+
+        if torch.cuda.is_available():
+            self.device_str = "cuda"
+        else:
+            self.device_str = "cpu"
+        self.device = torch.device(self.device_str)
+        use_auth_token = hf_login() if self.cfg.needs_auth else None
+        self.model = AutoModelForCausalLM.from_pretrained(
+            self.cfg.model_name,
+            torch_dtype=getattr(torch, self.cfg.compute_dtype),
+            use_auth_token=use_auth_token,
+            device_map=self.device_str,
+        )
+        self.tokenizer = AutoTokenizer.from_pretrained(self.cfg.model_name, use_auth_token=use_auth_token)
+        self.tokenizer.use_default_system_prompt = False
+        self.logger = logger
+
+    def chat(
+        self,
+        message: str,
+        system_prompt: str,
+        chat_history: Optional[List[Tuple[str, str]]] = None,
+        max_new_tokens: int = 1024,
+        temperature: float = 0.6,
+        top_p: float = 0.9,
+        top_k: int = 50,
+        repetition_penalty: float = 1.2,
+        num_beams: int = 1,
+    ) -> Iterable[str]:
+        """Chat with the model."""
+        conversation = []
+        if system_prompt:
+            conversation.append({"role": "system", "content": system_prompt})
```

Comment:
[nit] this is a bit hard to follow. Can we define a template with fields for conversation entries?